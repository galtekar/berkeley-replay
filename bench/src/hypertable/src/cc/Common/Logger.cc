/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include "Logger.h"

#include <log4cpp/Appender.hh>
#include <log4cpp/BasicLayout.hh>
#include <log4cpp/FileAppender.hh>
#include <log4cpp/Layout.hh>
#include <log4cpp/NDC.hh>
#include <log4cpp/OstreamAppender.hh>
#include <log4cpp/Priority.hh>

using namespace Hypertable;

namespace {

  /**
   * NoTimeLayout
   **/
  class  NoTimeLayout : public log4cpp::Layout {
  public:
    NoTimeLayout() { }
    virtual ~NoTimeLayout() { }
    virtual std::string format(const log4cpp::LoggingEvent& event) {
      std::ostringstream message;
      const std::string& priorityName = log4cpp::Priority::getPriorityName(event.priority);
      message << priorityName << " " << event.categoryName << " " << event.ndc << ": " << event.message << std::endl;
      return message.str();
    }
  };

}


log4cpp::Category *Logger::logger = 0;
bool Logger::ms_show_line_numbers = true;


void Logger::initialize(const char *name, log4cpp::Priority::Value priority) {
  log4cpp::Appender* appender = 
    new log4cpp::OstreamAppender("default", &std::cout);
  log4cpp::Layout* layout = new log4cpp::BasicLayout();
  appender->setLayout(layout);
  logger = &(log4cpp::Category::getInstance(std::string(name)));
  logger->addAppender(appender);
  logger->setAdditivity(false);
  logger->setPriority(priority);
}


void Logger::set_level(log4cpp::Priority::Value priority) {
  logger->setPriority(priority);
}

void Logger::suppress_line_numbers() {
  ms_show_line_numbers = false;
}


void Logger::set_test_mode(const char *name) {
  log4cpp::FileAppender *appender;

  ms_show_line_numbers = false;

  logger->removeAllAppenders();
  appender = new log4cpp::FileAppender((std::string)name, 1);
  appender->setLayout(new NoTimeLayout());
  logger->setAppender(appender);
}
