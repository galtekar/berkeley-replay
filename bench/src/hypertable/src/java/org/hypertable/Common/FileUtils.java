/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */


package org.hypertable.Common;

import java.io.InputStream;
import java.io.IOException;
import java.io.File;
import java.io.FileInputStream;

public class FileUtils {

    // Returns the contents of the file in a byte array.
    public static byte[] FileToBuffer(File file) throws IOException {
	InputStream is = new FileInputStream(file);
    
	// Get the size of the file
	long length = file.length();
    
	// You cannot create an array using a long type.
	// It needs to be an int type.
	// Before converting to an int type, check
	// to ensure that file is not larger than Integer.MAX_VALUE.
	if (length > Integer.MAX_VALUE) {
	    // File is too large
	}
    
	// Create the byte array to hold the data
	byte[] bytes = new byte[(int)length];
    
	// Read in the bytes
	int offset = 0;
	int numRead = 0;
	while (offset < bytes.length
	       && (numRead=is.read(bytes, offset, bytes.length-offset)) >= 0) {
	    offset += numRead;
	}
    
	// Ensure all the bytes have been read in
	if (offset < bytes.length) {
	    throw new IOException("Could not completely read file " + file.getName());
	}
    
	// Close the input stream and return bytes
	is.close();

	return bytes;
    }
}
