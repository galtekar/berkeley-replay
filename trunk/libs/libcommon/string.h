#pragma once

#ifndef __cplusplus

#include "debug.h"

/* ----- String with reference counting ----- */

typedef struct {
   char *strP;
   int count;
} String;

static INLINE String *
String_Alloc(size_t len)
{
   String *sP = malloc(sizeof(*sP));
   sP->strP = malloc(len);
   sP->count = 1;

   return sP;
}

static INLINE String *
String_Get(String *sP)
{
   sP->count++;

   return sP;
}

static INLINE void
String_Put(String *sP)
{
   sP->count--;

   if (sP->count == 0) {
      ASSERT(sP->strP);
      free(sP->strP);
      sP->strP = NULL;
      free(sP);
      sP = NULL;
   }
}

#endif
