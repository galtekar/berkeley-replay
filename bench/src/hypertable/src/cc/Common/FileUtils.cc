/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
#include <iomanip>
#include <iostream>
#include <sstream>

extern "C" {
#include <errno.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/uio.h>
#include <unistd.h>
}

#include <boost/shared_array.hpp>

#include "FileUtils.h"
#include "Logger.h"

using namespace Hypertable;
using namespace std;

/**
 */
ssize_t FileUtils::read(int fd, void *vptr, size_t n) {
  size_t nleft;
  ssize_t nread;
  char *ptr;

  ptr = (char *)vptr;
  nleft = n;
  while (nleft > 0) {
    if ( (nread = ::read(fd, ptr, nleft)) < 0) {
      if (errno == EINTR)
	nread = 0;/* and call read() again */
      else if (errno == EAGAIN)
	break;
      else
	return -1;
    } else if (nread == 0)
      break;/* EOF */

    nleft -= nread;
    ptr   += nread;
  }
  return n - nleft;
}

/**
 */
ssize_t FileUtils::pread(int fd, void *vptr, size_t n, off_t offset) {
  size_t nleft;
  ssize_t nread;
  char *ptr;

  ptr = (char *)vptr;
  nleft = n;
  while (nleft > 0) {
    if ( (nread = ::pread(fd, ptr, nleft, offset)) < 0) {
      if (errno == EINTR)
	nread = 0;/* and call read() again */
      else if (errno == EAGAIN)
	break;
      else
	return -1;
    } else if (nread == 0)
      break;/* EOF */

    nleft -= nread;
    ptr   += nread;
    offset += nread;
  }
  return n - nleft;
}

/**
 */
ssize_t FileUtils::write(int fd, const void *vptr, size_t n) {
  size_t nleft;
  ssize_t nwritten;
  const char *ptr;

  ptr = (const char *)vptr;
  nleft = n;
  while (nleft > 0) {
    if ((nwritten = ::write(fd, ptr, nleft)) <= 0) {
      if (errno == EINTR)
	nwritten = 0; /* and call write() again */
      if (errno == EAGAIN)
	break;
      else
	return -1; /* error */
    }

    nleft -= nwritten;
    ptr   += nwritten;
  }
  return n - nleft;
}

ssize_t FileUtils::writev(int fd, const struct iovec *vector, int count) {
  ssize_t nwritten;
  while ((nwritten = ::writev(fd, vector, count)) <= 0) {
    if (errno == EINTR)
      nwritten = 0; /* and call write() again */
    if (errno == EAGAIN) {
      nwritten = 0;
      break;
    }
    else
      return -1; /* error */
  }
  return nwritten;
}


ssize_t FileUtils::sendto(int fd, const void *vptr, size_t n, const struct sockaddr *to, socklen_t tolen) {
  size_t nleft;
  ssize_t nsent;
  const char *ptr;

  ptr = (const char *)vptr;
  nleft = n;
  while (nleft > 0) {
    if ((nsent = ::sendto(fd, ptr, nleft, 0, to, tolen)) <= 0) {
      if (errno == EINTR)
	nsent = 0; /* and call sendto() again */
      if (errno == EAGAIN || errno == ENOBUFS)
	break;
      else
	return -1; /* error */
    }

    nleft -= nsent;
    ptr   += nsent;
  }
  return n - nleft;
}

ssize_t FileUtils::recvfrom(int fd, void *vptr, size_t n, struct sockaddr *from, socklen_t *fromlen) {
  ssize_t nread;
  while (true) {
    if ( (nread = ::recvfrom(fd, vptr, n, 0, from, fromlen)) < 0) {
      if (errno != EINTR)
	break;
    }
    else
      break;
  }
  return nread;
}


/* flags are file status flags to turn on */
void FileUtils::set_flags(int fd, int flags) {
  int val;

  if ( (val = fcntl(fd, F_GETFL, 0)) < 0)
    cerr << "fcnt(F_GETFL) failed : " << strerror(errno) << endl;

  val |= flags;

  if (fcntl(fd, F_SETFL, val) < 0)
    cerr << "fcnt(F_SETFL) failed : " << strerror(errno) << endl;
}



/**
 */
char *FileUtils::file_to_buffer(const char *fname, off_t *lenp) {
  struct stat statbuf;
  int fd;

  *lenp = 0;

  if ((fd = open(fname, O_RDONLY)) < 0) {
    HT_ERRORF("open(\"%s\") failure - %s", fname,  strerror(errno));
    return 0;
  }

  if (fstat(fd, &statbuf) < 0) {
    HT_ERRORF("fstat(\"%s\") failure - %s", fname,  strerror(errno));
    return 0;
  }

  *lenp = statbuf.st_size;

  char *rbuf = new char [ *lenp + 1 ];

  ssize_t nread = FileUtils::read(fd, rbuf, *lenp);

  if (nread == (ssize_t)-1) {
    HT_ERRORF("read(\"%s\") failure - %s", fname,  strerror(errno));
    delete [] rbuf;
    *lenp = 0;
    return 0;
  }

  if (nread < *lenp) {
    HT_WARNF("short read (%d of %d bytes)", nread, *lenp);
    *lenp = nread;
  }

  rbuf[nread] = 0;

  return rbuf;
}


bool FileUtils::mkdirs(const char *dirname) {
  struct stat statbuf;
  boost::shared_array<char> tmpDirPtr(new char [ strlen(dirname) + 1 ]);
  char *tmpDir = tmpDirPtr.get();
  char *ptr = tmpDir+1;

  strcpy(tmpDir, dirname);
  
  while ((ptr = strchr(ptr, '/')) != 0) {
    *ptr = 0;
    if (stat(tmpDir, &statbuf) != 0) {
      if (errno == ENOENT) {
	if (mkdir(tmpDir, 0755) != 0) {
	  HT_ERRORF("Problem creating directory '%s' - %s", tmpDir, strerror(errno));
	  return false;
	}
      }
      else {
	HT_ERRORF("Problem stat'ing directory '%s' - %s", tmpDir, strerror(errno));
	return false;
      }
    }
    *ptr++ = '/';
  }

  if (stat(tmpDir, &statbuf) != 0) {
    if (errno == ENOENT) {
      if (mkdir(tmpDir, 0755) != 0) {
	HT_ERRORF("Problem creating directory '%s' - %s", tmpDir, strerror(errno));
	return false;
      }
    }
    else {
      HT_ERRORF("Problem stat'ing directory '%s' - %s", tmpDir, strerror(errno));
      return false;
    }
  }

  return true;
}


bool FileUtils::exists(const char *fname) {
  struct stat statbuf;
  if (stat(fname, &statbuf) != 0)
    return false;
  return true;
}

uint64_t FileUtils::size(const char *fname) {
  struct stat statbuf;
  if (stat(fname, &statbuf) != 0)
    return 0;
  return statbuf.st_size;
  
}


off_t FileUtils::length(const char *fname) {
  struct stat statbuf;
  if (stat(fname, &statbuf) != 0)
    return (off_t)-1;
  return statbuf.st_size;
}


int FileUtils::getxattr(const char *path, const char *name, void *value, size_t size) {
  std::string canonicalName;
  if (!strncmp(name, "user.", 5))
    canonicalName = name;
  else
    canonicalName = (std::string)"user." + name;
#if defined(__linux__)
  return ::getxattr(path, canonicalName.c_str(), value, size);
#elif defined(__APPLE__)
  return ::getxattr(path, canonicalName.c_str(), value, size, 0, 0);
#else
  ImplementMe;
#endif
}


int FileUtils::setxattr(const char *path, const char *name, const void *value, size_t size, int flags) {
  std::string canonicalName;
  if (!strncmp(name, "user.", 5))
    canonicalName = name;
  else
    canonicalName = (std::string)"user." + name;
#if defined(__linux__)
  return ::setxattr(path, canonicalName.c_str(), value, size, flags);
#elif defined(__APPLE__)
  return ::setxattr(path, canonicalName.c_str(), value, size, 0, flags);
#else
  ImplementMe;
#endif
}


int FileUtils::fgetxattr(int fd, const char *name, void *value, size_t size) {
  std::string canonicalName;
  if (!strncmp(name, "user.", 5))
    canonicalName = name;
  else
    canonicalName = (std::string)"user." + name;
#if defined(__linux__)
  return ::fgetxattr(fd, canonicalName.c_str(), value, size);
#elif defined(__APPLE__)
  return ::fgetxattr(fd, canonicalName.c_str(), value, size, 0, 0);
#else
  ImplementMe;
#endif
}


int FileUtils::fsetxattr(int fd, const char *name, const void *value, size_t size, int flags) {
  std::string canonicalName;
  if (!strncmp(name, "user.", 5))
    canonicalName = name;
  else
    canonicalName = (std::string)"user." + name;
#if defined(__linux__)
  return ::fsetxattr(fd, canonicalName.c_str(), value, size, flags);
#elif defined(__APPLE__)
  return ::fsetxattr(fd, canonicalName.c_str(), value, size, 0, flags);
#else
  ImplementMe;
#endif
}


int FileUtils::removexattr(const char *path, const char *name) {
  std::string canonicalName;
  if (!strncmp(name, "user.", 5))
    canonicalName = name;
  else
    canonicalName = (std::string)"user." + name;
#if defined(__linux__)
  return ::removexattr(path, canonicalName.c_str());
#elif defined(__APPLE__)
  return ::removexattr(path, canonicalName.c_str(), 0);
#else
  ImplementMe;
#endif
}

int FileUtils::fremovexattr(int fd, const char *name) {
  std::string canonicalName;
  if (!strncmp(name, "user.", 5))
    canonicalName = name;
  else
    canonicalName = (std::string)"user." + name;
#if defined(__linux__)
  return ::fremovexattr(fd, canonicalName.c_str());
#elif defined(__APPLE__)
  return ::fremovexattr(fd, canonicalName.c_str(), 0);
#else
  ImplementMe;
#endif

}
