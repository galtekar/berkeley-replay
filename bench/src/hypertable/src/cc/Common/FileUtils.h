/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
#ifndef HYPERTABLE_FDUTILS_H
#define HYPERTABLE_FDUTILS_H

extern "C" {
#include <fcntl.h>
#include <stdint.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/uio.h>
#include <sys/xattr.h>
#if defined(__linux__)
#include <attr/xattr.h>
#include <string.h>
#include <stdlib.h>
#endif
}

namespace Hypertable {

  class FileUtils {

  public:
    static ssize_t read(int fd, void *vptr, size_t n);
    static ssize_t pread(int fd, void *vptr, size_t n, off_t offset);
    static ssize_t write(int fd, const void *vptr, size_t n);
    static ssize_t writev(int fd, const struct iovec *vector, int count);
    static ssize_t sendto(int fd, const void *vptr, size_t n, const struct sockaddr *to, socklen_t tolen);
    static ssize_t recvfrom(int fd, void *vptr, size_t n, struct sockaddr *from, socklen_t *fromlen);
    static void set_flags(int fd, int flags);
    static char *file_to_buffer(const char *fname, off_t *lenp);
    static bool mkdirs(const char *dirname);
    static bool exists(const char *fname);
    static uint64_t size(const char *fname);
    static off_t length(const char *fname);
    static int getxattr(const char *path, const char *name, void *value, size_t size);
    static int fgetxattr(int fd, const char *name, void *value, size_t size);
    static int setxattr(const char *path, const char *name, const void *value, size_t size, int flags);
    static int fsetxattr(int fd, const char *name, const void *value, size_t size, int flags);
    static int removexattr(const char *path, const char *name);
    static int fremovexattr(int fd, const char *name);
  };

}

#endif // HYPERTABLE_FDUITLS_H

