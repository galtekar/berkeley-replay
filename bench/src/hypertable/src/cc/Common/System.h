/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifndef HYPERTABLE_SYSTEM_H
#define HYPERTABLE_SYSTEM_H

#include <string>
#include <string.h>

#include <boost/thread/mutex.hpp>

#define DUMP_CORE *((int *)0) = 1;

namespace Hypertable {

  class System {
    
  public:
    static void initialize(const char *argv0);

    static std::string installDir;
    static std::string executableName;
    
    static int get_processor_count();

  private:
    static boost::mutex ms_mutex;
  };

}

#endif // HYPERTABLE_SYSTEM_H
