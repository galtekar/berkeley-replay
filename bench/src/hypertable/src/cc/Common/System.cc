/**
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <cstdlib>
#include <cstring>
#include <iostream>
#include <string>

extern "C" {
#include <stdlib.h>
#include <sys/types.h>
#include <sys/sysctl.h>
#include <unistd.h>
}

#include "FileUtils.h"
#include "Logger.h"
#include "System.h"

using namespace Hypertable;
using namespace std;


string System::installDir;
string System::executableName;
boost::mutex System::ms_mutex;

void System::initialize(const char *argv0) {
  boost::mutex::scoped_lock lock(ms_mutex);
  const char *execPath = getenv("_");
  char cwd[1024];
  int offset;

  if (installDir != "")
    return;

  srand((unsigned)getpid());

  getcwd(cwd, 1024);
  strcat(cwd, "/");

  char *ptr = strrchr(argv0, '/');
  if (ptr == 0) {
    executableName = argv0;
#if 0
    std::string fname;
    char *path = getenv("PATH");
    char *last;
    execPath = 0;
    ptr = strtok_r(path, ":", &last);
    while (ptr) {
      fname = (std::string)ptr + "/" + executableName;
      if (FileUtils::exists(fname.c_str())) {
	execPath = fname.c_str();
	break;
      }
      ptr = strtok_r(0, ":", &last);      
    }
#endif
  }
  else
    executableName = ptr+1;

  offset = (execPath) ? strlen(execPath) - strlen(argv0) : -1;
  if (execPath && offset >= 0) {
    if (execPath[0] != '/') {
      installDir = cwd;
      installDir += execPath;
    }
    else
      installDir = execPath;
  }
  else {
    if (argv0[0] != '/') {
      installDir = cwd;
      installDir += argv0;
    }
    else
      installDir = argv0;
  }

  string::size_type pos = installDir.rfind("/bin/");
  if (pos == string::npos) {
    string::size_type pos = installDir.rfind("/");
    if (pos == string::npos)
      installDir = ".";
    else
      installDir.erase(pos);
    //cerr << "WARNING: " << executableName << " does not appear to have been run from an installation." << endl;
    //cerr << "WARNING: Using '" << installDir << "' as the installation directory" << endl;
  }
  else
    installDir.erase(pos);

  Logger::initialize(executableName.c_str());
}


int System::get_processor_count() {
#if defined(__APPLE__)
  int mib[2], ncpus;
  size_t len;

  mib[0] = CTL_HW;
  mib[1] = HW_NCPU;
  len = sizeof(ncpus);
  sysctl(mib, 2, &ncpus, &len, NULL, 0);
  return ncpus;
#elif defined(__linux__)
  return (int)sysconf(_SC_NPROCESSORS_ONLN);
#else
  ImplementMe;
#endif
}
