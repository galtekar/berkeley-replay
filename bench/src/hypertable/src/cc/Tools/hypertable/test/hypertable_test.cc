/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <cstdlib>
#include <iostream>
#include <string>

extern "C" {
#include <sys/types.h>
#include <unistd.h>
}

#include "Common/FileUtils.h"
#include "Common/Logger.h"
#include "Common/System.h"

using namespace Hypertable;
using namespace std;


namespace {
  const char *requiredFiles[] = {
    "./hypertable_test",
    "./hypertable.cfg",
    "./hypertable_test.hql",
    "./hypertable_test.golden",
    0
  };
}

int main(int argc, char **argv) {
  std::string commandStr;

  System::initialize(argv[0]);

  for (int i=0; requiredFiles[i]; i++) {
    if (!FileUtils::exists(requiredFiles[i])) {
      HT_ERRORF("Unable to find '%s'", requiredFiles[i]);
      return 1;
    }
  }

  /**
   *  hypertable_test
   */
  commandStr = (std::string)"./hypertable --no-prompt --config=hypertable.cfg < hypertable_test.hql > hypertable_test.output 2>&1";
  if (system(commandStr.c_str()) != 0)
    return 1;

  commandStr = (std::string)"diff hypertable_test.output hypertable_test.golden";
  if (system(commandStr.c_str()) != 0)
    return 1;

  return 0;
}
