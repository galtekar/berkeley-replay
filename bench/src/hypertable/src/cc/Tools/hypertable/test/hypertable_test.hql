DROP TABLE IF EXISTS hypertable;
CREATE TABLE hypertable (
apple,
banana
);
insert into hypertable VALUES ('2007-12-02 08:00:00', 'foo', 'apple:0', 'nothing'), ('2007-12-02 08:00:01', 'foo', 'apple:1', 'nothing'), ('2007-12-02 08:00:02', 'foo', 'apple:2', 'nothing');
insert into hypertable VALUES ('2007-12-02 08:00:03', 'foo', 'banana:0', 'nothing'), ('2007-12-02 08:00:04', 'foo', 'banana:1', 'nothing'), ('2007-12-02 08:00:05', 'bar', 'banana:2', 'nothing');
select * from hypertable display_timestamps;
delete "apple:1" from hypertable where row = 'foo' timestamp '2007-12-02 08:00:01';
select * from hypertable display_timestamps;
delete banana from hypertable where row = 'foo';
select * from hypertable display_timestamps;
insert into hypertable VALUES ('how', 'apple:0', 'nothing'), ('how', 'apple:1', 'nothing'), ('how', 'apple:2', 'nothing');
insert into hypertable VALUES ('now', 'banana:0', 'nothing'), ('now', 'banana:1', 'nothing'), ('now', 'banana:2', 'nothing');
insert into hypertable VALUES ('2007-12-02 08:00:00', 'lowrey', 'apple:0', 'nothing'), ('2007-12-02 08:00:00', 'season', 'apple:1', 'nothing'), ('2007-12-02 08:00:00', 'salt', 'apple:2', 'nothing');
insert into hypertable VALUES ('2028-02-17 08:00:01', 'lowrey', 'apple:0', 'nothing');
insert into hypertable VALUES ('2028-02-17 08:00:00', 'season', 'apple:1', 'nothing');
drop table if exists Pages;	
create table Pages ( "refer-url", "http-code", timestamp, rowkey, ACCESS GROUP default ( "refer-url", "http-code", timestamp, rowkey ) );
insert into Pages VALUES ('2008-01-28 22:00:03',  "calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern", 'http-code', '200' );
select "http-code" from Pages where ROW = "calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern" display_timestamps;
delete * from Pages where ROW = "calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern" TIMESTAMP '2008-01-28 22:00:10';
select "http-code" from Pages where ROW = "calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern" display_timestamps;
quit
