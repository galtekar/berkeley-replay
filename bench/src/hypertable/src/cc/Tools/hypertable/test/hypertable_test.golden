Welcome to the HQL command interpreter.

Type 'help;' or '\h' for help. Type '\c' to clear the buffer.

DROP TABLE IF EXISTS hypertable;
CREATE TABLE hypertable (
apple,
banana
);
insert into hypertable VALUES ('2007-12-02 08:00:00', 'foo', 'apple:0', 'nothing'), ('2007-12-02 08:00:01', 'foo', 'apple:1', 'nothing'), ('2007-12-02 08:00:02', 'foo', 'apple:2', 'nothing');
insert into hypertable VALUES ('2007-12-02 08:00:03', 'foo', 'banana:0', 'nothing'), ('2007-12-02 08:00:04', 'foo', 'banana:1', 'nothing'), ('2007-12-02 08:00:05', 'bar', 'banana:2', 'nothing');
select * from hypertable display_timestamps;
2007-12-02 08:00:05.000000000	bar	banana:2	nothing
2007-12-02 08:00:03.000000000	foo	banana:0	nothing
2007-12-02 08:00:04.000000000	foo	banana:1	nothing
2007-12-02 08:00:00.000000000	foo	apple:0	nothing
2007-12-02 08:00:01.000000000	foo	apple:1	nothing
2007-12-02 08:00:02.000000000	foo	apple:2	nothing
delete "apple:1" from hypertable where row = 'foo' timestamp '2007-12-02 08:00:01';
delete: row='foo' family=2 qualifier='1' ts=1196582401000000001
select * from hypertable display_timestamps;
2007-12-02 08:00:05.000000000	bar	banana:2	nothing
2007-12-02 08:00:03.000000000	foo	banana:0	nothing
2007-12-02 08:00:04.000000000	foo	banana:1	nothing
2007-12-02 08:00:00.000000000	foo	apple:0	nothing
2007-12-02 08:00:02.000000000	foo	apple:2	nothing
delete banana from hypertable where row = 'foo';
delete: row='foo' family=1 ts=0
select * from hypertable display_timestamps;
2007-12-02 08:00:05.000000000	bar	banana:2	nothing
2007-12-02 08:00:00.000000000	foo	apple:0	nothing
2007-12-02 08:00:02.000000000	foo	apple:2	nothing
insert into hypertable VALUES ('how', 'apple:0', 'nothing'), ('how', 'apple:1', 'nothing'), ('how', 'apple:2', 'nothing');
insert into hypertable VALUES ('now', 'banana:0', 'nothing'), ('now', 'banana:1', 'nothing'), ('now', 'banana:2', 'nothing');
insert into hypertable VALUES ('2007-12-02 08:00:00', 'lowrey', 'apple:0', 'nothing'), ('2007-12-02 08:00:00', 'season', 'apple:1', 'nothing'), ('2007-12-02 08:00:00', 'salt', 'apple:2', 'nothing');
Error: Problem sending updates (table=HYPERTABLE) - RANGE SERVER supplied timestamp is not strictly increasing
insert into hypertable VALUES ('2028-02-17 08:00:01', 'lowrey', 'apple:0', 'nothing');
insert into hypertable VALUES ('2028-02-17 08:00:00', 'season', 'apple:1', 'nothing');
Error: Problem sending updates (table=HYPERTABLE) - RANGE SERVER supplied timestamp is not strictly increasing
drop table if exists Pages;
create table Pages ( "refer-url", "http-code", timestamp, rowkey, ACCESS GROUP default ( "refer-url", "http-code", timestamp, rowkey ) );
insert into Pages VALUES ('2008-01-28 22:00:03',  "calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern", 'http-code', '200' );
select "http-code" from Pages where ROW = "calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern" display_timestamps;
2008-01-28 22:00:03.000000000	calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern	http-code	200
delete * from Pages where ROW = "calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern" TIMESTAMP '2008-01-28 22:00:10';
delete: row='calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern' family=0 ts=1201557610000000001
select "http-code" from Pages where ROW = "calendar.boston.com/abington-ma/venues/show/457680-the-cellar-tavern" display_timestamps;
